clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -189,7 +185,41 @@ func (d *ClusterDeployer) Create(cluster *clusterv1.Cluster, machines []*cluster
 		}
 	}
 
-	glog.Infof("Done provisioning cluster. You can now access your cluster with kubectl --kubeconfig %v", d.kubeconfigOutput)
+	glog.Infof("Done provisioning cluster. You can now access your cluster with kubectl --kubeconfig %v", kubeconfigOutput)
+
+	return nil
+}
+
+func (d *ClusterDeployer) Delete(internalClient ClusterClient) error {
+	glog.Info("Creating external cluster")
+	externalClient, cleanupExternalCluster, err := d.createExternalCluster()
+	defer cleanupExternalCluster()
+	if err != nil {
+		return fmt.Errorf("could not create external cluster: %v", err)
+	}
+	defer closeClient(externalClient, "external")
+
+	glog.Info("Applying Cluster API stack to external cluster")
+	if err = d.applyClusterAPIStack(externalClient); err != nil {
+		return fmt.Errorf("unable to apply cluster api stack to external cluster: %v", err)
+	}
+
+	glog.Info("Deleting Cluster API Provider Components from internal cluster")
+	if err = internalClient.Delete(d.providerComponents); err != nil {
+		glog.Infof("Error while shutting down provider components on internal cluster: %v", err)
+		glog.Infof("Continuing with a best effort delete")
+	}
+
+	glog.Info("Copying objects from internal cluster to external cluster")
+	if err = pivot(internalClient, externalClient); err != nil {
+		return fmt.Errorf("unable to copy objects from internal to external cluster: %v", err)
+	}
+
+	glog.Info("Deleting objects from external cluster")
+	if err = deleteObjects(externalClient); err != nil {
+		return fmt.Errorf("unable to finish deleting objects in external cluster, resources may have been leaked: %v", err)
```

Comment:
It's hard to build a generalized line of text here because the underlying resources are sort of provider specific. For example, in GCP when the cluster deletion fails it means that you may have 'leaked' a firewall rule. However, that would likely not be true for a vsphere implementation.

Do you have any ideas on how to improve the messaging?

Reply:
I don't have a good solution right now, but perhaps we could make a generic statement ("VMs, firewall rules, LB rules etc") for now?

Reply:
I'm not sure that is actually more clear because the resources in question are extremely specific to a given provider, for example, firewall rule is a GCP concept, in AWS there is "security groups", VMs might be safe, but in AWS land those would be called instances. When talking about load balancers, within AWS there are at least 3 different types of load balancers with different APIs, etc, (application, network, and classic). 

I see two options:

1. Add another method to the ProviderDeployer interface. This is the interface that is specific to the provider / cloud. This method could be something like this `GetDeleteMachineErrorMessage(...) string` and there would be one for Clusters, Machines, MachineSets, and MachineDeployments. This method would return a provider specific message of the kinds of resources that can be leaked.

2. Change the message to this, substituting 'google' or 'aws' or the like for the provider name:
```
fmt.Errorf("unable to finish deleting objects in external cluster, the associated resources specific to the %v provider may have been leaked: %v", providerName, err)
```